STM32/FLASH_readwrite/Src/FLASH_readwrite.c

```diff
@@ -2,139 +2,158 @@
-// The flash is used for program data and this must NOT be cleared. Putting data in
-// linker script will ensure that program data will not be cleared when writing
-// calibration data to the flash. To access these variables the syntax below is used.
-extern uint32_t _FlashSector; // Variable defined in ld linker script.
-extern uint32_t _FlashAddr;   // Variable defined in ld linker script.
-#define FLASH_SECTOR ((uint32_t) &_FlashSector)
-#define FLASH_ADDR   ((uint32_t) &_FlashAddr)
-
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-// Write data to FLASH_ADDR+indx directly. This method does not
-// ensure any data integrity as data validation is performed.
-void writeToFlash(uint32_t indx, uint32_t size, uint8_t *data)
+static uint32_t getFlashSector(uint32_t address)
 {
-    __HAL_RCC_WWDG_CLK_DISABLE();
+    uint32_t address_offset = address - START_SECTOR_0;
+    uint32_t address_sectors_start[6] = {START_SECTOR_0, START_SECTOR_1, START_SECTOR_2, START_SECTOR_3,
+                                         START_SECTOR_4, START_SECTOR_5};
+
+    uint32_t address_sectors_end[6] = {END_SECTOR_0, END_SECTOR_1, END_SECTOR_2, END_SECTOR_3,
+                                       END_SECTOR_4, END_SECTOR_5};                                         
+    for (int i = 0; i < 6; i++)
+    {
+        if (address_offset >= address_sectors_start[i]-START_SECTOR_0 &&
+            address_offset <= address_sectors_end[i]-START_SECTOR_0)
+        {
+            return i;
+        }
+    }
+    // This should never occur - return non-existing sector.
+    return 10;
+}
 
+/*!
+** @brief Store data in FLASH memory.
+** @note Does not store a CRC so data integrity verification is not possible when reading the 
+**       data at later stage.
+*/
+void writeToFlash(uint32_t flash_address, uint8_t *data, uint32_t size)
+{
+    __HAL_RCC_WWDG_CLK_DISABLE();
+    uint32_t flash_sector = getFlashSector(flash_address);
     // Erase the sector before write
     HAL_FLASH_Unlock();
-    FLASH_Erase_Sector(FLASH_SECTOR, FLASH_VOLTAGE_RANGE_3);
+    FLASH_Erase_Sector(flash_sector, FLASH_VOLTAGE_RANGE_3);
     HAL_FLASH_Lock();
 
     // Write to sector
     HAL_FLASH_Unlock();
 
     for(uint32_t i=0; i<size; i++)
     {
-        HAL_FLASH_Program(FLASH_TYPEPROGRAM_BYTE, FLASH_ADDR+indx+i , data[i]);
+        HAL_FLASH_Program(FLASH_TYPEPROGRAM_BYTE, flash_address+i , data[i]);
     }
     HAL_FLASH_Lock();
     __HAL_RCC_WWDG_CLK_ENABLE();
 }
 
-// Read from flash without checking data integrity if CRC module
-// is not enabled in project.
-void readFromFlash(uint32_t indx, uint32_t size, uint8_t *data)
+/*!
+** @brief Read data from FLASH memory at flash_adress.
+*/
+void readFromFlash(uint32_t flash_address, uint8_t *data, uint32_t size)
 {
     for(uint32_t i=0; i<size; i++)
     {
-        *(data + i) = *( (uint8_t *)(FLASH_ADDR+indx+i) );
+        *(data + i) = *( (uint8_t *)(flash_address+i) );
     }
```

Comment:
I believe (although am happy to be proved wrong) that memcpy is a superior way (in terms of performance) of executing this function.
https://cplusplus.com/reference/cstring/memcpy/